src/dodal/devices/utils.py

```diff
@@ -12,3 +18,71 @@ def epics_signal_put_wait(pv_name: str, wait: float = 1.0) -> EpicsSignal:
         EpicsSignal: An EpicsSignal that will wait for a callback.
     """
     return Component(EpicsSignal, pv_name, put_complete=True, write_timeout=wait)
+
+
+def wrap_and_do_funcs(unwrapped_funcs, timeout=60):
```

Comment:
@DominicOram Have been thinking about a cleaner implementation, how about something like this?
It's not completely right because some of the tests don't like it, but I think recursion may be the way to go:

```python
def wrap_and_do_funcs(
    unwrapped_funcs: list[Callable[[], Status]],
    timeout: float = 60.0,
    previous_status: Optional[Status] = None,
) -> Status:
    func, other_funcs = unwrapped_funcs[-1], unwrapped_funcs[1:-1]

    def wrapper(previous_status: Optional[Status] = None) -> Status:
        if previous_status is not None and previous_status.exception() is not None:
            # If there is an exception, return the status immediately, ends recursion
            # so no further statuses are evaluated
            LOGGER.error(
                f"{previous_status} has failed with error {previous_status.exception()}"
            )
            return previous_status

        LOGGER.info(f"Doing {func.__name__}")
        next_status = func()
        if len(other_funcs) > 0:
            # Queue up the next function to run when this status completes.
            # Do not queue up more functions if there aren't any, ends recursion
            next_status.add_callback(lambda: wrap_and_do_funcs(other_funcs, timeout))
        return next_status

    return wrapper(previous_status)

```

Reply:
Yes, I feel like we can do it nicer with recursion. I think that might require a bit more thought so let's do a beamline test first @olliesilvester then we'll try and refactor to use recursion